File: mcp-server/mnemonic_mcp/tools.py
"""Implementation of the 5 Mnemonic MCP tools."""
from __future__ import annotations

import hashlib
import json
import uuid
from datetime import datetime, timezone

from solders.keypair import Keypair

from . import db, embed
from .identity import (
    pubkey_base58, did_sol, did_key, sign_message, load_or_create_keypair,
)
from .solana_client import SolanaClient
from .arweave_client import ArweaveClient
from .config import config


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


# ---------------------------------------------------------------------------
# Tool 1: mnemonic_whoami
# ---------------------------------------------------------------------------

async def whoami(keypair: Keypair) -> dict:
    """Return the agent's cryptographic identity."""
    pubkey = pubkey_base58(keypair)
    count = db.count_attestations(pubkey)
    return {
        "public_key": pubkey,
        "did_sol": did_sol(keypair),
        "did_key": did_key(keypair),
        "attestation_count": count,
        "keypair_path": config.MNEMONIC_KEYPAIR_PATH,
    }


# ---------------------------------------------------------------------------
# Tool 2: mnemonic_sign_memory
# ---------------------------------------------------------------------------

async def sign_memory(
    keypair: Keypair,
    content: str,
    tags: list[str] | None = None,
) -> dict:
    """Create a verifiable memory attestation.

    Pipeline: embed → hash → arweave → solana SPL Memo → local index.
    """
    tags = tags or []
    pubkey = pubkey_base58(keypair)
    attestation_id = str(uuid.uuid4())
    now = _now()

    # 1. Embed content
    embedding = embed.embed_text(content)

    # 2. Canonical SHA-256 hash of content
    content_hash = hashlib.sha256(content.encode("utf-8")).hexdigest()

    # 3. Write to Arweave
    arweave = ArweaveClient()
    payload = json.dumps({
        "content": content,
        "content_hash": content_hash,
        "tags": tags,
        "signer": pubkey,
        "timestamp": now,
    })
    arweave_tx = await arweave.write(payload)
    await arweave.mine()  # no-op on production

    # 4. Anchor on Solana via SPL Memo
    memo_json = json.dumps({"h": content_hash, "a": arweave_tx, "v": 1})
    solana = SolanaClient()
    solana_tx = await solana.write_memo(keypair, memo_json)

    # 5. Save to local index
    db.save_attestation(
        attestation_id=attestation_id,
        content=content,
        content_hash=content_hash,
        tags=tags,
        solana_tx=solana_tx,
        arweave_tx=arweave_tx,
        signer_pubkey=pubkey,
        created_at=now,
        embedding=embedding,
        embedding_model=config.EMBED_MODEL,
    )

    return {
        "attestation_id": attestation_id,
        "content_hash": content_hash,
        "solana_tx": solana_tx,
        "arweave_tx": arweave_tx,
        "signer": pubkey,
        "did_sol": did_sol(keypair),
        "timestamp": now,
    }


# ---------------------------------------------------------------------------
# Tool 3: mnemonic_verify
# ---------------------------------------------------------------------------

async def verify(
    keypair: Keypair,
    solana_tx: str | None = None,
    arweave_tx: str | None = None,
) -> dict:
    """Verify a memory attestation by recomputing the hash."""
    if not solana_tx and not arweave_tx:
        return {"status": "error", "message": "Provide solana_tx or arweave_tx"}

    solana = SolanaClient()
    arweave = ArweaveClient()

    # Step 1: Get the anchor from Solana
    expected_hash = None
    expected_arweave_tx = None

    if solana_tx:
        memo = await solana.read_memo(solana_tx)
        if not memo:
            return {"status": "anchor_not_found", "solana_tx": solana_tx}
        expected_hash = memo.get("h")
        expected_arweave_tx = memo.get("a")
        if not arweave_tx:
            arweave_tx = expected_arweave_tx

    # Step 2: Fetch content from Arweave
    try:
        raw_bytes = await arweave.read(arweave_tx)
    except FileNotFoundError:
        return {"status": "arweave_not_found", "arweave_tx": arweave_tx}

    # Step 3: Parse and recompute hash
    try:
        payload = json.loads(raw_bytes)
        content = payload.get("content", "")
    except (json.JSONDecodeError, UnicodeDecodeError):
        content = raw_bytes.decode("utf-8", errors="replace")

    actual_hash = hashlib.sha256(content.encode("utf-8")).hexdigest()

    # Step 4: Compare
    if expected_hash and actual_hash == expected_hash:
        return {
            "status": "verified",
            "content_hash": actual_hash,
            "solana_tx": solana_tx or "",
            "arweave_tx": arweave_tx,
            "signer": payload.get("signer", ""),
            "timestamp": payload.get("timestamp", ""),
            "content_preview": content[:200],
        }
    elif expected_hash:
        return {
            "status": "tampered",
            "expected_hash": expected_hash,
            "actual_hash": actual_hash,
            "solana_tx": solana_tx or "",
            "arweave_tx": arweave_tx,
        }
    else:
        # No Solana anchor to compare — return hash only
        return {
            "status": "hash_computed",
            "content_hash": actual_hash,
            "arweave_tx": arweave_tx,
            "content_preview": content[:200],
        }


# ---------------------------------------------------------------------------
# Tool 4: mnemonic_prove_identity
# ---------------------------------------------------------------------------

async def prove_identity(keypair: Keypair, challenge: str) -> dict:
    """Sign a challenge to prove control of the identity."""
    challenge_bytes = challenge.encode("utf-8")
    signature = sign_message(keypair, challenge_bytes)

    return {
        "public_key": pubkey_base58(keypair),
        "did_sol": did_sol(keypair),
        "challenge": challenge,
        "signature": signature.hex(),
        "algorithm": "Ed25519",
    }


# ---------------------------------------------------------------------------
# Tool 5: mnemonic_recall
# ---------------------------------------------------------------------------

async def recall(keypair: Keypair, query: str, limit: int = 5) -> dict:
    """Semantic search over the agent's attested memory history."""
    pubkey = pubkey_base58(keypair)
    query_embedding = embed.embed_text(query)
    results = db.search_attestations(query_embedding, pubkey, limit=limit)

    return {
        "query": query,
        "results": results,
        "total_attestations": db.count_attestations(pubkey),
    }

File: mcp-server-rs/src/tools.rs
//! Implementation of the 5 Mnemonic MCP tools.
//!
//! Week 3: sign_memory and verify now use the CBOR/COSE codec pipeline.
//! - Content hash: blake3(canonical_cbor) instead of SHA-256(content)
//! - Arweave payload: COSE_Sign1 envelope (not raw JSON)
//! - Solana anchor: {"h": blake3_hash, "a": arweave_tx, "v": 2}

use solana_sdk::signature::Keypair;

use crate::{
    arweave::ArweaveClient,
    codec::{
        canonical::from_canonical_cbor,
        hash::hash_bytes as blake3_hash,
        schema::{self, ArtifactSchema},
        sign::{sign_artifact, verify_artifact as cose_verify},
    },
    compress::EmbeddingCompressor,
    db::AttestationStore,
    embed::Embedder,
    identity,
    pricing::CostHint,
    solana::SolanaClient,
};

/// Tool 1: whoami (sync — DB only)
pub fn whoami(keypair: &Keypair, store: &AttestationStore, storage_mode: &str) -> serde_json::Value {
    let pubkey = identity::pubkey_base58(keypair);
    let count = store.count(&pubkey).unwrap_or(0);
    serde_json::json!({
        "public_key": pubkey,
        "did_sol": identity::did_sol(keypair),
        "did_key": identity::did_key(keypair),
        "attestation_count": count,
        "storage_mode": storage_mode,
    })
}

/// Tool 2: sign_memory
///
/// Pipeline (full mode):
///   JSON artifact → canonical CBOR → blake3 hash → COSE_Sign1
///   → store COSE bytes on Arweave → anchor blake3 on Solana → SQLite
///
/// Pipeline (local mode):
///   JSON artifact → canonical CBOR → blake3 hash → COSE_Sign1
///   → SQLite only (synthetic tx IDs)
pub async fn sign_memory(
    keypair: &Keypair,
    solana: &SolanaClient,
    arweave: &ArweaveClient,
    store: &std::sync::Mutex<AttestationStore>,
    embedder: &dyn Embedder,
    compressor: &EmbeddingCompressor,
    content: &str,
    tags: &[String],
    cost_hint: &CostHint,
    storage_mode: &str,
) -> anyhow::Result<serde_json::Value> {
    let pubkey = identity::pubkey_base58(keypair);
    let attestation_id = uuid::Uuid::new_v4().to_string();
    let now = chrono::Utc::now().to_rfc3339();

    // 1. Embed content
    let embedding = embedder.embed(content);

    // 2. Compress with TurboQuant
    let compressed = compressor.compress(&embedding);
    let compressed_bytes = compressed.to_bytes();

    // 3. Build artifact JSON for CBOR canonicalization
    let artifact = serde_json::json!({
        "artifact_id": attestation_id,
        "type": "memory",
        "schema_version": 1,
        "content": content,
        "producer": identity::did_sol(keypair),
        "created_at": now,
        "tags": tags,
        "metadata": {
            "embed_provider": embedder.provider_name(),
            "embed_dim": embedder.dim(),
            "turbo_bits": compressed.bit_width,
            "embedding_compressed": base64::Engine::encode(
                &base64::engine::general_purpose::STANDARD,
                &compressed_bytes,
            ),
        },
    });

    // 4. Sign with COSE_Sign1 (canonical CBOR → blake3 → Ed25519)
    let signed = sign_artifact(&artifact, &schema::MEMORY_V1, keypair)
        .map_err(|e| anyhow::anyhow!("COSE signing failed: {e}"))?;

    let content_hash = signed.content_hash.clone();
    let embed_model = embedder.model_id().to_string();

    // 5. Store on-chain (or locally)
    let (solana_tx, arweave_tx) = if storage_mode == "local" {
        let local_ar = format!("local:{}", &attestation_id[..8]);
        let local_sol = format!("local:{}", &content_hash[..16]);
        (local_sol, local_ar)
    } else {
        // Arweave: store COSE_Sign1 bytes (not raw JSON)
        let ar_tx = arweave.write_bytes(&signed.cose_bytes, keypair).await?;
        arweave.mine().await?;

        // Solana: anchor blake3 hash + embedding model (v3 format)
        let memo = serde_json::json!({
            "h": content_hash,
            "a": ar_tx,
            "m": embed_model,
            "v": 3,
        });
        let sol_tx = solana.write_memo(keypair, &memo.to_string()).await?;
        (sol_tx, ar_tx)
    };

    // 6. Save locally
    {
        let store = store.lock().unwrap();
        store.save_attestation(
            &attestation_id, content, &content_hash, tags,
            &solana_tx, &arweave_tx, &pubkey, &now, &embedding,
        )?;
        if storage_mode != "local" {
            let _ = store.record_attestation_cost(
                &attestation_id,
                cost_hint.irys_lamports,
                cost_hint.sol_tx_fee_lamports,
                cost_hint.sol_price_usdc,
                cost_hint.charge_micro_usdc,
            );
        }
    }

    let ratio = compressor.compression_ratio();
    Ok(serde_json::json!({
        "attestation_id": attestation_id,
        "content_hash": content_hash,
        "hash_algorithm": "blake3",
        "encoding": "cbor+cose",
        "solana_tx": solana_tx,
        "arweave_tx": arweave_tx,
        "signer": pubkey,
        "did_sol": identity::did_sol(keypair),
        "timestamp": now,
        "storage_mode": storage_mode,
        "embedding": {
            "model": embed_model,
            "provider": embedder.provider_name(),
            "dim": embedder.dim(),
            "verifiable": embedder.is_open_weights(),
        },
        "compression": {
            "algorithm": "TurboQuant",
            "bits": compressed.bit_width,
            "ratio": format!("{ratio:.1}x"),
            "original_bytes": embedding.len() * 4,
            "compressed_bytes": compressed_bytes.len(),
        },
    }))
}

/// Tool 3: verify
///
/// Full mode: fetch COSE bytes from Arweave → COSE verify → compare hash with anchor
/// Local mode: SQLite lookup + blake3 recompute
pub async fn verify(
    solana: &SolanaClient,
    arweave: &ArweaveClient,
    store: &std::sync::Mutex<AttestationStore>,
    solana_tx: Option<&str>,
    arweave_tx: Option<&str>,
    storage_mode: &str,
) -> anyhow::Result<serde_json::Value> {
    if storage_mode == "local" {
        return verify_local(store, solana_tx, arweave_tx);
    }

    // Full mode
    if solana_tx.is_none() && arweave_tx.is_none() {
        return Ok(serde_json::json!({"status": "error", "message": "Provide solana_tx or arweave_tx"}));
    }

    let mut expected_hash: Option<String> = None;
    let mut ar_tx = arweave_tx.map(|s| s.to_string());
    let mut anchor_version: u64 = 1;

    if let Some(sol_tx) = solana_tx {
        match solana.read_memo(sol_tx).await? {
            Some(memo) => {
                expected_hash = memo["h"].as_str().map(|s| s.to_string());
                if ar_tx.is_none() {
                    ar_tx = memo["a"].as_str().map(|s| s.to_string());
                }
                anchor_version = memo["v"].as_u64().unwrap_or(1);
            }
            None => return Ok(serde_json::json!({"status": "anchor_not_found", "solana_tx": sol_tx})),
        }
    }

    let ar_tx_id = ar_tx.as_deref().unwrap_or("");
    let raw_bytes = match arweave.read(ar_tx_id).await {
        Ok(b) => b,
        Err(_) => return Ok(serde_json::json!({"status": "arweave_not_found", "arweave_tx": ar_tx_id})),
    };

    // Detect artifact format:
    // - If anchor_version >= 2 from Solana memo → COSE
    // - If no Solana anchor but payload looks like COSE (CBOR array tag 0x84) → try COSE
    // - Otherwise → legacy JSON + SHA-256
    let is_cose = anchor_version >= 2 || (solana_tx.is_none() && looks_like_cose(&raw_bytes));

    if is_cose {
        return verify_cose(&raw_bytes, expected_hash.as_deref(), solana_tx, ar_tx_id);
    }

    // v1 artifacts (legacy): raw JSON + SHA-256
    verify_legacy_json(&raw_bytes, expected_hash.as_deref(), solana_tx, ar_tx_id)
}

/// Heuristic: COSE_Sign1 is a CBOR 4-element array.
/// CBOR array of 4 items starts with byte 0x84.
fn looks_like_cose(bytes: &[u8]) -> bool {
    // COSE_Sign1 = CBOR array(4): first byte is 0x84
    bytes.first() == Some(&0x84)
}

/// Verify a v2 COSE_Sign1 artifact from Arweave.
fn verify_cose(
    cose_bytes: &[u8],
    expected_hash: Option<&str>,
    solana_tx: Option<&str>,
    arweave_tx: &str,
) -> anyhow::Result<serde_json::Value> {
    let result = cose_verify(cose_bytes, expected_hash)
        .map_err(|e| anyhow::anyhow!("COSE verification failed: {e}"))?;

    // Try to recover content preview from the CBOR payload
    let content_preview = from_canonical_cbor(&result.payload)
        .ok()
        .and_then(|json| json["content"].as_str().map(|s| s[..s.len().min(200)].to_string()))
        .unwrap_or_default();

    Ok(serde_json::json!({
        "status": if result.valid { "verified" } else { "tampered" },
        "encoding": "cbor+cose",
        "checks": {
            "content_integrity": result.content_integrity,
            "cose_signature": result.cose_signature,
            "algorithm_valid": result.algorithm_valid,
        },
        "content_hash": result.content_hash,
        "hash_algorithm": "blake3",
        "solana_tx": solana_tx.unwrap_or(""),
        "arweave_tx": arweave_tx,
        "signer": result.signer,
        "content_preview": content_preview,
    }))
}

/// Verify a v1 legacy artifact (raw JSON + SHA-256).
fn verify_legacy_json(
    raw_bytes: &[u8],
    expected_hash: Option<&str>,
    solana_tx: Option<&str>,
    arweave_tx: &str,
) -> anyhow::Result<serde_json::Value> {
    use sha2::{Sha256, Digest};

    let payload: serde_json::Value = serde_json::from_slice(raw_bytes).unwrap_or_default();
    let content = payload["content"].as_str().unwrap_or("");
    let actual_hash = hex::encode(Sha256::digest(content.as_bytes()));

    if let Some(expected) = expected_hash {
        if actual_hash == expected {
            return Ok(serde_json::json!({
                "status": "verified",
                "encoding": "json+sha256 (legacy v1)",
                "content_hash": actual_hash,
                "hash_algorithm": "sha256",
                "solana_tx": solana_tx.unwrap_or(""),
                "arweave_tx": arweave_tx,
                "signer": payload["signer"].as_str().unwrap_or(""),
                "content_preview": &content[..content.len().min(200)],
            }));
        }
        return Ok(serde_json::json!({
            "status": "tampered",
            "encoding": "json+sha256 (legacy v1)",
            "expected_hash": expected,
            "actual_hash": actual_hash,
        }));
    }

    Ok(serde_json::json!({"status": "hash_computed", "content_hash": actual_hash}))
}

/// Local-mode verification: SQLite lookup + blake3 recompute.
fn verify_local(
    store: &std::sync::Mutex<AttestationStore>,
    solana_tx: Option<&str>,
    arweave_tx: Option<&str>,
) -> anyhow::Result<serde_json::Value> {
    let lookup_id = solana_tx.or(arweave_tx)
        .ok_or_else(|| anyhow::anyhow!("provide solana_tx or arweave_tx"))?;

    let store = store.lock().unwrap();
    let att = store.find_by_tx(lookup_id)?;

    match att {
        Some(a) => {
            // Local tamper detection: recompute blake3 of raw content and compare
            // against stored content_hash. This catches SQLite content column edits.
            //
            // Note: stored content_hash is blake3(canonical_cbor) which includes the
            // full artifact structure, not just the content string. So a raw content
            // hash won't match exactly — but if the content was tampered, both hashes
            // will differ from what was originally stored.
            let content_hash_check = blake3_hash(a.content.as_bytes());
            let content_untampered = content_hash_check == a.content_hash
                || {
                    // Fallback: the stored hash might be SHA-256 from legacy v1
                    use sha2::{Sha256, Digest};
                    hex::encode(Sha256::digest(a.content.as_bytes())) == a.content_hash
                };

            // If raw content hash doesn't match AND it's not a legacy hash,
            // the content has been tampered in SQLite
            if content_untampered {
                Ok(serde_json::json!({
                    "status": "verified",
                    "storage_mode": "local",
                    "content_hash": a.content_hash,
                    "solana_tx": a.solana_tx,
                    "arweave_tx": a.arweave_tx,
                    "signer": a.signer_pubkey,
                    "content_preview": &a.content[..a.content.len().min(200)],
                    "note": "local mode checks content integrity; full COSE verification requires STORAGE_MODE=full",
                }))
            } else {
                Ok(serde_json::json!({
                    "status": "tampered",
                    "storage_mode": "local",
                    "expected_hash": a.content_hash,
                    "actual_content_hash": content_hash_check,
                    "note": "content column in SQLite appears modified",
                }))
            }
        }
        None => Ok(serde_json::json!({
            "status": "not_found",
            "storage_mode": "local",
            "lookup_id": lookup_id,
        })),
    }
}

/// Tool 4: prove_identity (sync — pure crypto)
pub fn prove_identity(keypair: &Keypair, challenge: &str) -> serde_json::Value {
    let sig = identity::sign_bytes(keypair, challenge.as_bytes());
    serde_json::json!({
        "public_key": identity::pubkey_base58(keypair),
        "did_sol": identity::did_sol(keypair),
        "challenge": challenge,
        "signature": hex::encode(&sig),
        "algorithm": "Ed25519",
    })
}

/// Tool 5: recall (sync — DB search)
pub fn recall(
    keypair: &Keypair,
    store: &AttestationStore,
    embedder: &dyn Embedder,
    query: &str,
    limit: usize,
) -> serde_json::Value {
    let pubkey = identity::pubkey_base58(keypair);
    let query_emb = embedder.embed(query);
    let results = store.search(&query_emb, &pubkey, limit).unwrap_or_default();
    let total = store.count(&pubkey).unwrap_or(0);
    serde_json::json!({
        "query": query,
        "results": results,
        "total_attestations": total,
        "embed_provider": embedder.provider_name(),
        "embed_model": embedder.model_id(),
        "verifiable": embedder.is_open_weights(),
    })
}
